[ONEROSTER-46] Add OneRoster service reference docs (#470)
Ed-Fi-Alliance-OSS/ed-fi-alliance-oss.github.io: docs/reference/11-oneroster/configuration/environment-variables.md

# Environment Variables

The OneRoster® Node service reads configuration from environment
variables, typically supplied via a `.env` file at the application
root. The sample `.env.example` in the service repository lists every
variable. The tables below group them by concern and note defaults and
required-or-optional status.

The service connects to the ODS API's `EdFi_Admin` database — not to
an ODS database directly. At request time, it looks up which ODS
instance to query from claims on the inbound JWT and resolves the ODS
connection string from `EdFi_Admin.OdsInstances`. See [OAuth and
JWT](./oauth-and-jwt.md) for the claim shapes the service expects.

## Minimum viable configuration

A minimum `.env` for a native PostgreSQL deployment against a
single-tenant `EdFi_Admin`:

```env
DB_TYPE=postgres

> CONNECTION_CONFIG={"adminConnection":"host=localhost;port=5432;database=EdFi_Admin;username=postgres;password=P@ssw0rd"}

Reviewer: Remove this to avoid the risk of documenting / normalizing using administrator account or using easy to guess password or use a template CONNECTION_CONFIG=CONNECTION_CONFIG={"adminConnection":"host=localhost;port=5432;database=EdFi_Admin;username=<your-username>;password=<your-password>"}

Reviewer 2: Done in `6b8a01b` — replaced `username=postgres;password=P@ssw0rd` with `<your-username>` / `<your-password>` placeholders in the         
  minimum-viable-config example.

| `API_BASE_PATH` | _(empty)_ | Set when the service is hosted under a virtual directory (for example, `/oneroster` behind IIS). Used to generate deterministic discovery URLs. |
| `NODE_ENV` | `dev` | Set to `prod` (or empty) for production. Influences logging verbosity. |

## Connection and tenancy

The service takes its database engine from `DB_TYPE` and one or two
admin-database connection blobs in JSON form. ODS connection strings
themselves are _not_ supplied here — they live encrypted inside
`EdFi_Admin.OdsInstances`, and the service decrypts them at request
time using `ODS_CONNECTION_STRING_ENCRYPTION_KEY`.

### Engine

| Variable | Default | Notes |
| --- | --- | --- |
| `DB_TYPE` | `postgres` | `postgres` or `mssql`. Applies to both the admin-database connection and the ODS connections resolved through it. |

### Admin database (single-tenant, default)

| Variable | Default | Notes |
| --- | --- | --- |
| `CONNECTION_CONFIG` | _(none)_ | Required in single-tenant mode. JSON with an `adminConnection` key whose value is an `EdFi_Admin` connection string for the configured `DB_TYPE`. Ignored when `MULTITENANCY_ENABLED=true`. |

> | `ODS_CONNECTION_STRING_ENCRYPTION_KEY` | _(none)_ | Required. Base64-encoded AES key used to decrypt the ODS connection strings stored in `EdFi_Admin.OdsInstances`. **Must match** the ODS API's `ApiSettings:OdsConnectionStringEncryptionKey`. Generate with `openssl rand -base64 32`. Without it, the service cannot connect to any ODS. |

Reviewer: Remove instruction to generate. The instruction should be to copy the setting from ODS API's setting. 
```suggestion
| `ODS_CONNECTION_STRING_ENCRYPTION_KEY` | _(none)_ | Required. Base64-encoded AES key used to decrypt the ODS connection strings stored in `EdFi_Admin.OdsInstances`. **Must match** the ODS API's `ApiSettings:OdsConnectionStringEncryptionKey`. Without this key, the service cannot connect to any ODS. |
```

Reviewer 2: Done in `4c6b383` — accepted the suggestion. Dropped the `openssl rand` hint and tightened "Without it" → "Without this key" so it reads  as "copy from the ODS API setting" rather than "generate fresh".

PostgreSQL example:

```env

> CONNECTION_CONFIG={"adminConnection":"host=localhost;port=5432;database=EdFi_Admin;username=postgres;password=P@ssw0rd"}

Reviewer: Remove this to avoid the risk of documenting / normalizing using administrator account or using easy to guess password or use a template `CONNECTION_CONFIG=CONNECTION_CONFIG={"adminConnection":"host=localhost;port=5432;database=EdFi_Admin;username=<your-username>;password=<your-password>"}`

Author: Done in `6b8a01b` — replaced with `<your-username>` / `<your-password>` placeholders.

```

Microsoft SQL Server example:

```env

> CONNECTION_CONFIG={"adminConnection":"server=localhost;database=EdFi_Admin;user id=sa;password=P@ssw0rd;encrypt=false;TrustServerCertificate=true"}

Reviewer: Remove this to avoid the risk of documenting / normalizing using administrator account or using easy to guess password or use a template `CONNECTION_CONFIG={"adminConnection":"server=localhost;database=EdFi_Admin;user id=<your-username>;password=<your-password>;}`

Author: Done in `6b8a01b` — replaced `user id=sa;password=P@ssw0rd` with `<your-username>` / `<your-password>` placeholders.

```

### Multi-tenant mode

| Variable | Default | Notes |
| --- | --- | --- |
| `MULTITENANCY_ENABLED` | `false` | Set to `true` to enable multi-tenant routing. Must match the ODS API's multi-tenancy setting. |
| `TENANTS_CONNECTION_CONFIG` | _(none)_ | Required when `MULTITENANCY_ENABLED=true`. JSON object mapping tenant identifier to an `{adminConnection: "..."}` entry. Tenant identifiers are matched case-insensitively against the `:tenantId` route segment and the JWT's `tenantId` claim. |

PostgreSQL example:

```env
MULTITENANCY_ENABLED=true

> TENANTS_CONNECTION_CONFIG={"Tenant1":{"adminConnection":"host=localhost;port=5432;database=EdFi_Admin_Tenant1;username=postgres;password=pass1"},"Tenant2":{"adminConnection":"host=localhost;port=5432;database=EdFi_Admin_Tenant2;username=postgres;password=pass2"}}

Reviewer: Same as above, avoid the risk of documenting / normalizing using administrator account or using easy to guess password

Author: Done in `6b8a01b` — applied the same `<your-username>` / `<your-password>` placeholder pattern to both tenant entries.

including the JWT claims used to pick the authorized ODS instance.

| Variable | Default | Notes |
| --- | --- | --- |
| `OAUTH2_ISSUERBASEURL` | _(none)_ | Required. Base URL of the OAuth 2.0 issuer (typically your Ed-Fi ODS / API's `/oauth/` endpoint). The `/.well-known/jwks.json` path is resolved against this when PEM verification is not configured. |
| `OAUTH2_AUDIENCE` | _(none)_ | Required. Expected `aud` claim on inbound JWTs. |
| `OAUTH2_TOKENSIGNINGALG` | `RS256` | JWT signing algorithm. |
| `OAUTH2_PUBLIC_KEY_PEM` | _(empty)_ | If set, the service verifies JWTs with this PEM public key instead of fetching JWKS. Use `\n` for line breaks when storing in `.env`. Must match the ODS API's `Security:Jwt:SigningKey:PublicKey`. |

## CORS, rate limiting, and proxy

See [CORS, rate limiting, and proxy](./cors-rate-limit-proxy.md) for
the behavior details.

| Variable | Default | Notes |
| --- | --- | --- |
| `CORS_ORIGINS` | `http://localhost:3000` | Comma-separated allowed origins. Leave empty to allow all (not recommended in production). |
| `RATE_LIMIT_WINDOW_MS` | `60000` | Rate-limit window in milliseconds (`express-rate-limit`). |

> | `RATE_LIMIT_MAX_REQUESTS` | `60` | Maximum requests per window per IP. The bundled `.env.example` sets this to `100`. The service code uses `60` as the fallback when the variable is unset. |

Reviewer: Both env.example and code default to 100 now
```suggestion
| `RATE_LIMIT_MAX_REQUESTS` | `60` | Maximum requests per window per IP.|
```

Reviewer 2: Done in `11f19b2` — set to `100`, dropped the (incorrect) prose claim that the code-level fallback was `60`. Treating the `60` in your suggestion block as a typo: `edfi-oneroster` main has `RATE_LIMIT_MAX_REQUESTS=100` in `.env.example` and `parseInt(...) || 100` in
  `src/app.js`. Flag if I read that wrong.
